truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java

```diff
@@ -60,10 +63,17 @@ public CachedDispatchNode(
         this.next = next;
     }
 
-    public void replaceSendingChild() {
+    public void replaceSendingFrame() {
         CompilerAsserts.neverPartOfCompilation("Dispatch nodes should not be altered after compilation.");
-        if (!needsCallerFrame) {
-            needsCallerFrame = true;
+        if (!sendsFrame) {
+            sendsFrame = true;
+        }
+    }
+
+    public void replaceSendingCallerFrame(CallerFrameAccess access) {
+        CompilerAsserts.neverPartOfCompilation("Dispatch nodes should not be altered after compilation.");
```

Comment:
You're right. The ReadCallerFrameNode should not try and change things if it has been compiled, The dispatch nodes should revert to the interpreter and invalidate.